File: http-servers/typescript/bun-elysia/src/routes/params.ts
import { Elysia } from "elysia";
import { DEFAULT_LIMIT, MAX_FILE_BYTES, NULL_BYTE, SNIFF_LEN } from "../consts/defaults";
import {
  FILE_NOT_FOUND,
  FILE_NOT_TEXT,
  FILE_SIZE_EXCEEDS,
  INVALID_FORM_DATA,
  INVALID_JSON_BODY,
  INVALID_MULTIPART,
  makeError,
  ONLY_TEXT_PLAIN
} from "../consts/errors";

export const paramsRouter = new Elysia();

paramsRouter.get("/search", ({ query }) => {
  const q = query.q?.trim() || "none";
  const limitStr = query.limit;
  const limitNum = Number(limitStr);
  const limit = Number.isSafeInteger(limitNum) ? limitNum : DEFAULT_LIMIT;
  return { search: q, limit };
});

paramsRouter.get("/url/:dynamic", ({ params: { dynamic } }) => {
  return { dynamic };
});

paramsRouter.get("/header", ({ headers }) => {
  const header = headers["x-custom-header"]?.trim() || "none";
  return { header };
});

paramsRouter.post("/body", ({ body, set }) => {
  if (typeof body !== "object" || body === null || Array.isArray(body)) {
    set.status = 400;
    return makeError(INVALID_JSON_BODY, "expected a JSON object");
  }
  return { body };
});

paramsRouter.get("/cookie", ({ cookie }) => {
  const fooValue = cookie.foo?.value;
  const cookieVal = typeof fooValue === "string" ? fooValue.trim() || "none" : "none";

  cookie.bar.value = "12345";
  cookie.bar.maxAge = 10;
  cookie.bar.httpOnly = true;
  cookie.bar.path = "/";

  return { cookie: cookieVal };
});

paramsRouter.post("/form", ({ body, set, headers }) => {
  const contentType = headers["content-type"]?.toLowerCase() ?? "";
  if (!contentType.startsWith("application/x-www-form-urlencoded") && !contentType.startsWith("multipart/form-data")) {
    set.status = 400;
    return makeError(
      INVALID_FORM_DATA,
      "expected content-type: application/x-www-form-urlencoded or multipart/form-data"
    );
  }

  if (typeof body !== "object" || body === null || Array.isArray(body)) {
    set.status = 400;
    return makeError(INVALID_FORM_DATA, "expected form fields");
  }

  const form = body as Record<string, unknown>;

  const name = typeof form.name === "string" && form.name.trim() !== "" ? form.name.trim() : "none";

  const ageStr = typeof form.age === "string" && form.age.trim() !== "" ? form.age.trim() : "0";
  const ageNum = Number(ageStr);
  const age = Number.isSafeInteger(ageNum) ? ageNum : 0;

  return { name, age };
});

paramsRouter.post("/file", async ({ body, set, headers }) => {
  const contentType = headers["content-type"]?.toLowerCase() ?? "";
  if (!contentType.startsWith("multipart/form-data")) {
    set.status = 400;
    return makeError(INVALID_MULTIPART, "expected content-type: multipart/form-data");
  }

  const form = body as Record<string, unknown>;
  const file = form.file;

  if (!(file instanceof File)) {
    set.status = 400;
    return makeError(FILE_NOT_FOUND, "no file field named 'file' in form data");
  }
  if (!file.type || !file.type.startsWith("text/plain")) {
    set.status = 415;
    return makeError(ONLY_TEXT_PLAIN, `received mimetype: ${file.type || "unknown"}`);
  }
  if (file.size > MAX_FILE_BYTES) {
    set.status = 413;
    return makeError(FILE_SIZE_EXCEEDS, `file size ${file.size} exceeds limit ${MAX_FILE_BYTES}`);
  }

  const buffer = await file.arrayBuffer();
  const data = new Uint8Array(buffer);
  if (data.length > MAX_FILE_BYTES) {
    set.status = 413;
    return makeError(FILE_SIZE_EXCEEDS, `file size ${data.length} exceeds limit ${MAX_FILE_BYTES}`);
  }

  const head = data.slice(0, SNIFF_LEN);
  if (head.includes(NULL_BYTE)) {
    set.status = 415;
    return makeError(FILE_NOT_TEXT, "file contains null bytes in header");
  }

  if (data.includes(NULL_BYTE)) {
    set.status = 415;
    return makeError(FILE_NOT_TEXT, "file contains null bytes");
  }

  let content: string;
  try {
    content = new TextDecoder("utf-8", { fatal: true }).decode(data);
  } catch {
    set.status = 415;
    return makeError(FILE_NOT_TEXT, "file is not valid UTF-8");
  }

  return {
    filename: file.name,
    size: data.length,
    content
  };
});

File: http-servers/python/fastapi/src/routes/params.py
from typing import Any

from fastapi import APIRouter, Body, Cookie, File, Header, Request, Response, UploadFile
from fastapi.exceptions import HTTPException

from src.consts.defaults import DEFAULT_LIMIT, MAX_FILE_BYTES, NULL_BYTE, SAFE_INT_LIMIT, SNIFF_LEN
from src.consts.errors import (
    EXPECTED_FORM_CONTENT_TYPE,
    EXPECTED_MULTIPART_CONTENT_TYPE,
    FILE_NOT_FOUND,
    FILE_NOT_TEXT,
    FILE_SIZE_EXCEEDS,
    INVALID_FORM_DATA,
    INVALID_JSON_BODY,
    INVALID_MULTIPART,
    ONLY_TEXT_PLAIN,
    make_error,
)


params_router = APIRouter()


def _strip_or(value: str | None, default: str) -> str:
    stripped = value.strip() if value else ""
    return stripped or default


def _parse_safe_int(value: str | None, default: int) -> int:
    if value is None or "." in value:
        return default
    try:
        num = int(value)
        if -SAFE_INT_LIMIT <= num <= SAFE_INT_LIMIT:
            return num
    except ValueError:
        pass
    return default


@params_router.get("/search")
async def search_params(q: str | None = None, limit: str | None = None):
    return {"search": _strip_or(q, "none"), "limit": _parse_safe_int(limit, DEFAULT_LIMIT)}


@params_router.get("/url/{dynamic}")
async def url_params(dynamic: str):
    return {"dynamic": dynamic}


@params_router.get("/header")
async def header_params(
    header: str | None = Header(alias="X-Custom-Header", default=None),
):
    return {"header": _strip_or(header, "none")}


@params_router.post("/body")
async def body_params(body: Any = Body()):
    if not isinstance(body, dict):
        raise HTTPException(status_code=400, detail=make_error(INVALID_JSON_BODY, "expected a JSON object"))
    return {"body": body}


@params_router.get("/cookie")
async def cookie_params(
    response: Response,
    foo: str | None = Cookie(default=None),
):
    response.set_cookie(key="bar", value="12345", max_age=10, httponly=True, path="/")
    return {"cookie": _strip_or(foo, "none")}


@params_router.post("/form")
async def form_params(request: Request):
    content_type = request.headers.get("content-type", "").lower()
    if not (
        content_type.startswith("application/x-www-form-urlencoded") or content_type.startswith("multipart/form-data")
    ):
        raise HTTPException(
            status_code=400,
            detail=make_error(INVALID_FORM_DATA, EXPECTED_FORM_CONTENT_TYPE),
        )

    try:
        form = await request.form()
    except Exception as e:
        raise HTTPException(
            status_code=400, detail=make_error(INVALID_FORM_DATA, str(e) or "failed to parse form body")
        )

    name_val = form.get("name")
    name = _strip_or(name_val if isinstance(name_val, str) else None, "none")

    age_val = form.get("age")
    age = _parse_safe_int(age_val if isinstance(age_val, str) and age_val.strip() else None, 0)

    return {"name": name, "age": age}


@params_router.post("/file")
async def file_params(request: Request, file: UploadFile | None = File(default=None)):
    content_type = request.headers.get("content-type", "").lower()
    if not content_type.startswith("multipart/form-data"):
        raise HTTPException(status_code=400, detail=make_error(INVALID_MULTIPART, EXPECTED_MULTIPART_CONTENT_TYPE))

    if file is None:
        raise HTTPException(
            status_code=400, detail=make_error(FILE_NOT_FOUND, "no file field named 'file' in form data")
        )

    if not file.content_type or not file.content_type.startswith("text/plain"):
        raise HTTPException(
            status_code=415, detail=make_error(ONLY_TEXT_PLAIN, f"received mimetype: {file.content_type or 'unknown'}")
        )

    data = await file.read(MAX_FILE_BYTES + 1)
    if len(data) > MAX_FILE_BYTES:
        raise HTTPException(
            status_code=413,
            detail=make_error(FILE_SIZE_EXCEEDS, f"file size {len(data)} exceeds limit {MAX_FILE_BYTES}"),
        )

    head = data[:SNIFF_LEN]
    if NULL_BYTE in head:
        raise HTTPException(
            status_code=415,
            detail=make_error(FILE_NOT_TEXT, "file contains null bytes in header"),
        )

    if NULL_BYTE in data:
        raise HTTPException(
            status_code=415,
            detail=make_error(FILE_NOT_TEXT, "file contains null bytes"),
        )

    try:
        content = data.decode("utf-8")
    except UnicodeDecodeError:
        raise HTTPException(
            status_code=415,
            detail=make_error(FILE_NOT_TEXT, "file is not valid UTF-8"),
        )

    return {
        "filename": file.filename,
        "size": len(data),
        "content": content,
    }
